src/GSoft.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenHttpMessageHandler.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System.Net;
+using System.Net.Http.Headers;
+using Microsoft.Extensions.Http;
+using Polly;
+
+namespace GSoft.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+/// <summary>
+/// HttpMessageHandler injected into the consumer's HttpClient.
+/// Adds HTTP Authentication headers with the access token retrieved from the client credentials grant flow.
+/// Retries the consumer's HTTP request only once using Polly with a new token if the previous token was expired or revoked.
+/// </summary>
+internal sealed class ClientCredentialsTokenHttpMessageHandler : PolicyHttpMessageHandler
+{
+    private const string RetryCountContextKey = "RetryCount";
+
+    // An expired or revoked token could result in an unauthorized HTTP error that we must retry only once with a new token
+    private static readonly IAsyncPolicy<HttpResponseMessage> RetryUnauthorizedResponseOnceAsyncPolicy = Policy<HttpResponseMessage>
+        .HandleResult(response => response.StatusCode == HttpStatusCode.Unauthorized)
+        .RetryAsync((_, retryCount, context) => context[RetryCountContextKey] = retryCount);
+
+    private readonly IClientCredentialsTokenManagementService _tokenManagementService;
+    private readonly string _clientName;
+
+    public ClientCredentialsTokenHttpMessageHandler(IClientCredentialsTokenManagementService tokenManagementService, string clientName)
+        : base(RetryUnauthorizedResponseOnceAsyncPolicy)
+    {
+        this._tokenManagementService = tokenManagementService;
+        this._clientName = clientName;
+    }
+
+    protected override async Task<HttpResponseMessage> SendCoreAsync(HttpRequestMessage request, Context context, CancellationToken cancellationToken)
+    {
+        // Retry with a new token if the previous attempt was unauthorized
+        var cachingBehavior = context.ContainsKey(RetryCountContextKey) ? CachingBehavior.ForceRefresh : CachingBehavior.PreferCache;
+        await this.SetTokenAsync(request, cachingBehavior, cancellationToken).ConfigureAwait(false);
+        return await base.SendCoreAsync(request, context, cancellationToken).ConfigureAwait(false);
+    }
+
+    private async Task SetTokenAsync(HttpRequestMessage request, CachingBehavior cachingBehavior, CancellationToken cancellationToken)
+    {
+        var token = await this._tokenManagementService.GetAccessTokenAsync(this._clientName, cachingBehavior, cancellationToken).ConfigureAwait(false);
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token.AccessToken);
```

Comment:
I think the MS library already include a string definition for the Bearer type authentication. It could be preferential to use it when referencing `Bearer`

Reply:
Actually there is no public constant exposed in the .NET SDK for `Bearer`. I've already done some research back then.

* Nothing in the .NET SDK, ([code search here](https://github.com/search?q=repo%3Adotnet%2Fruntime+%22Bearer%22&type=code))
* In the ASP.NET Core framework (which we don't reference here) there's only `JwtBearerDefaults.AuthenticationScheme` but its purpose is different so it's not a valid candidate ([code search here](https://github.com/search?q=repo%3Adotnet%2Faspnetcore+%22%5C%22Bearer%22+language%3AC%23&type=code&l=C%23&p=1))

Even in MSAL they actually hardcode it too: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.53.0/src/client/Microsoft.Identity.Client/AuthenticationResult.cs#L110